### PR TITLE
DEV-2360  | Support back behavior from Stripe Payment form to checkout

### PR DESCRIPTION
**Pull request by benjaminEwhite**
Please fill out each section even if it's just with "N/A"

#### What's this PR do?

Adds a "Back" button to the Stripe Checkout form. When clicked, the user is taken back to the initial check out form, and a request is made to the API that:

- Sets the (NRE) contribution status to "cancelled"
- Cancels the payment intent (if one-time) or subscription (if recurring) on Stripe

Here are some of the more note-worthy code changes:

- `OneTimePaymentViewSet` and `SubscriptionPaymentViewSet` have been refactored and combined into a single viewset that can handle both cases: `PaymentViewset`.  
- Relatedly, there is now a single URL for this view at `/payments`.
- `PaymentViewset` has a `destroy` method which sets contribution status to 'cancelled' and deletes respective Stripe payment intent or subscription.
- Adds @testing-library/cypress as dev spa dependency. This can be used to have more a11y-centric Cypress tests.
- Adds a new `BackButton` component in line with styles for revised page editor.

#### Why are we doing this? How does it help us?

Better contributor experience.

#### How should this be manually tested? Please include detailed step-by-step instructions.

1. Go to a live donation page in review app. Fill out the first form. You'll be taken to the Stripe Checkout form.
2. Look in the Django admin for your contribution. It should have a status of "processing".  While in Django admin, grab the provider_payment_id (if one time) or provider_subscription_id (if recurring).
3. Go to Stripe dashboard and locate either the payment intent (if one time) or subscription (if recurring). 
4. Now, back in the donation page, click the "Back" button. 
5. You should get taken back to the initial donation page form. The values you originally chose should still be filled in.
6. Back in Django admin, refresh the contribution. Its status should now be "cancelled".
7. Back in Stripe, the payment intent / subscription should no longer appear (having been deleted).
#### Have automated unit tests been added? If not, why?

Yes. On both front- and back-end.  

#### How should this change be communicated to end users?

N/A

#### Are there any smells or added technical debt to note?

No

#### Has this been documented? If so, where?

No

#### What are the relevant tickets? Add a link to any relevant ones.

https://news-revenue-hub.atlassian.net/browse/DEV-2360

#### Do any changes need to be made before deployment to production (adding environment variables, for example)?

No

#### Are there next steps? If so, what? Have tickets been opened for them? List next-step tickets here:

No

### CODE REVIEW

**Review by nrh-cklimas — CHANGES_REQUESTED** [review]
Just small changes here.

Our prettier precommit hook isn't working on TypeScript (yet)--see the fix on Slack, so I would either plug that in or run prettier manually. I noticed there's are also some extraneous line breaks here and there too.

File: spa/package.json [hunk 1]
```diff
@@ -152,6 +152,7 @@
     "@storybook/preset-create-react-app": "^4.1.2",
     "@storybook/react": "^6.5.10",
     "@storybook/testing-library": "^0.0.13",
+    "@testing-library/cypress": "^8.0.3",
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply spa/package.json:155]: Yeah, I think we can incrementally/opportunistically start supplanting getByTestId.

File: spa/src/components/common/Button/BackButton/BackButton.test.tsx [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+describe('BackButton', () => {
+
+  it('should render back button with default text', () => {
+    render(<BackButton {...props} />);
+    const button = screen.getByRole('button', {name: DEFAULT_BACK_BUTTON_TEXT});
+    expect(button).toBeEnabled();
+  });
+
+  it('should have an onClick callback function', () => {
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply spa/src/components/common/Button/BackButton/BackButton.test.tsx]: b9244b6625af

File: spa/src/components/common/Button/BackButton/BackButton.tsx [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+import { Button, Icon, Span} from './BackButton.styled';
+
+const BackButtonPropTypes = {
+  onClick: PropTypes.func.isRequired,
+  text: PropTypes.string,
+};
+
+interface BackButtonProps extends InferProps<typeof BackButtonPropTypes> {
+  onClick: (event: React.MouseEvent) => void;
+  text?: string
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply spa/src/components/common/Button/BackButton/BackButton.tsx]: 8fcb29836

File: spa/src/components/donationPage/DonationPage.js [hunk 4]
```diff
@@ -33,15 +33,22 @@ import { CONTRIBUTION_INTERVALS } from 'constants/contributionIntervals';
...
   // for page data (that happens in parent context)
-  return axios.post(apiEndpoint, paymentData, { headers: { [CSRF_HEADER]: csrftoken } }).then(({ data }) => data);
+  return axios
+    .post(AUTHORIZE_STRIPE_PAYMENT_ROUTE, paymentData, { headers: { [CSRF_HEADER]: csrftoken } })
+    .then(({ data }) => data);
+}
+
+function cancelPayment(paymentId, csrftoken) {
+  const endpoint = `${AUTHORIZE_STRIPE_PAYMENT_ROUTE}${paymentId}/`;
+  // we manually set the XCSRFToken value in header here. This is an unauthed endpoint
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply spa/src/components/donationPage/DonationPage.js]: Strictly speaking, it's `X-CSRFTOKEN' that gets set.

7d824f0c14

File: spa/src/components/donationPage/DonationPage.js [hunk 5]
```diff
@@ -258,7 +258,11 @@ function DonationPage({ page, live = false }) {
         stripeClientSecret,
         mailingCountry,
         setMailingCountry,
-        paymentSubmitButtonText
+        paymentSubmitButtonText,
+        cancelPayment: () => {
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply spa/src/components/donationPage/DonationPage.js:273]: d36cbf7115eca5

File: spa/src/components/common/Button/BackButton/BackButton.tsx [hunk 6]
```diff
@@ -0,0 +1,25 @@
+import React from 'react';
+import PropTypes, { InferProps } from 'prop-types';
+
+import { Button, Icon, Span} from './BackButton.styled';
+
+const BackButtonPropTypes = {
+  onClick: PropTypes.func.isRequired,
+  text: PropTypes.string,
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply spa/src/components/common/Button/BackButton/BackButton.tsx:8]: I'm inclined to leave as is, and generalize when we have > 1 case.

File: spa/src/components/donationPage/DonationPage.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (nrh-cklimas) [reply spa/src/components/donationPage/DonationPage.js]: Ah ok. I honestly thought it was a typo because X is next is to the C key.

**Review by nrh-cklimas — APPROVED** [review]

File: apps/contributions/views.py [hunk 7]
```diff
@@ -200,6 +203,31 @@ def get_serializer_context(self):
         context.update({"request": self.request})
         return context
 
+    def destroy(self, request, *args, **kwargs):
+        contribution = self.get_object()
+        if contribution.status != ContributionStatus.PROCESSING:
+            return Response(status=status.HTTP_409_CONFLICT)
```
Comment (njharman) [inline apps/contributions/views.py:209]: Should we log or return more information (for future debugging/diagnosis?  Such as the contribution status that was not PROCESSING. And, for log, identifying "ids" of contribution, contributor, DP, etc.

File: apps/contributions/views.py [hunk 8]
```diff
@@ -174,24 +181,20 @@ def process_stripe_webhook_view(request):
...
-    permission_classes = []
-    serializer_class = serializers.CreateRecurringPaymentSerializer
-    lookup_field = "provider_client_secret_id"
+    def get_serializer_class(self):
+        if (interval := self.request.data["interval"]) == ContributionInterval.ONE_TIME:
+            return serializers.CreateOneTimePaymentSerializer
+        elif interval in (ContributionInterval.MONTHLY.value, ContributionInterval.YEARLY.value):
+            return serializers.CreateRecurringPaymentSerializer
+        else:
+            raise ValidationError({"interval": "The provided value for interval is not permittted"})
```
Comment (njharman) [inline apps/contributions/views.py:197]: This "failure" path is not test covered.

File: apps/contributions/views.py [hunk 9]
```diff
@@ -200,6 +203,31 @@ def get_serializer_context(self):
...
+                stripe.PaymentIntent.cancel(
+                    contribution.provider_payment_id,
+                    stripe_account=contribution.donation_page.revenue_program.stripe_account_id,
+                )
+            if contribution.interval in (ContributionInterval.MONTHLY.value, ContributionInterval.YEARLY.value):
+                stripe.Subscription.delete(
+                    contribution.provider_subscription_id,
+                    stripe_account=contribution.donation_page.revenue_program.stripe_account_id,
+                )
+        except StripeError:
```
Comment (njharman) [inline apps/contributions/views.py:223]: This "failure" path is not test covered.

**Review by njharman — CHANGES_REQUESTED** [review]
Only reviewed Python code.

Lacking test coverage for unhappy paths and one suggestion on logging/error reporting. Otherwise looks good.

File: apps/contributions/views.py (hunk 7 above)
Comment (benjaminEwhite) [reply apps/contributions/views.py:220]: I don't think we should return more info to client, but agree re: logging, as this is not something the user (or SPA more generally) can/should fix.

e03a1341d

File: apps/contributions/views.py (hunk 8 above)
Comment (benjaminEwhite) [reply apps/contributions/views.py:197]: 1ec861552

Comment (benjaminEwhite) [reply apps/contributions/views.py]: See 1ec861552

File: apps/contributions/views.py (hunk 9 above)
Comment (benjaminEwhite) [reply apps/contributions/views.py:232]: See 838b60c36

**Review by x110dc — APPROVED** [review]

**Pull request merged**